File: src/infinicore/ops/random_sample/random_sample.cc
#include "infinicore/ops/random_sample.hpp"

namespace infinicore::op {

common::OpDispatcher<RandomSample::schema> &RandomSample::dispatcher() {
    static common::OpDispatcher<RandomSample::schema> dispatcher_;
    return dispatcher_;
};

void RandomSample::execute(
    Tensor indices, Tensor logits,
    float random_val, float topp, int topk, float temperature) {
    dispatcher().lookup(context::getDevice().getType())(
        indices, logits, random_val, topp, topk, temperature);
}

Tensor random_sample(
    Tensor logits,
    float random_val,
    float topp,
    int topk,
    float temperature) {
    auto indices = Tensor::empty({}, DataType::I32, logits->device());
    random_sample_(indices, logits, random_val, topp, topk, temperature);
    return indices;
}

void random_sample_(
    Tensor indices,
    Tensor logits,
    float random_val,
    float topp,
    int topk,
    float temperature) {
    RandomSample::execute(indices, logits, random_val, topp, topk, temperature);
}

} // namespace infinicore::op

File: python/infinicore/ops/random_sample.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def random_sample(logits, random_val, topp, topk, temperature, *, out=None):
    if out is None:
        return Tensor(
            _infinicore.random_sample(
                logits._underlying, random_val, topp, topk, temperature
            )
        )

    _infinicore.random_sample_(
        out._underlying,
        logits._underlying,
        random_val,
        topp,
        topk,
        temperature,
    )


